vm/vmm/vmm.go

```diff
@@ -0,0 +1,422 @@
+		sshport:   22,
+	}
+	closeInst := inst
+	defer func() {
+		if closeInst != nil {
+			closeInst.Close()
+		}
+	}()
+
+	if err := inst.Boot(); err != nil {
+		return nil, err
+	}
+
+	closeInst = nil
+	return inst, nil
+}
+
+func (pool *Pool) nextID() int {
+	pool.mu.Lock()
+	pool.count++
+	id := pool.count
+	pool.mu.Unlock()
+	return id
+}
+
+func (inst *instance) Boot() error {
+	name := fmt.Sprintf("syzkaller-%d", inst.imageID)
+	mem := fmt.Sprintf("%dM", inst.cfg.Mem)
+	startArgs := []string{
+		"start", name,
+		"-t", inst.cfg.Template,
+		"-b", inst.cfg.Kernel,
+		"-d", inst.image,
+		"-m", mem,
+	}
+	startOut, err := inst.vmctl(startArgs...)
+	if err != nil {
+		return fmt.Errorf("start failed: %v: %s", err, string(startOut))
+	}
+	if inst.debug {
+		log.Logf(0, "start output: %s", string(startOut))
+	}
+
+	inst.vmID, err = parseID(string(startOut))
+	if err != nil {
+		return err
+	}
+	inst.sshhost = fmt.Sprintf("100.64.%d.3", inst.vmID)
+
+	var tee io.Writer
+	if inst.debug {
+		tee = os.Stdout
+	}
+	inst.merger = vmimpl.NewOutputMerger(tee)
+
+	inst.console()
+
+	var bootOutput []byte
+	bootOutputStop := make(chan bool)
+	go func() {
+		for {
+			select {
+			case out := <-inst.merger.Output:
+				bootOutput = append(bootOutput, out...)
+			case <-bootOutputStop:
+				bootOutputStop <- true
+				return
+			}
+		}
+	}()
+
+	done := func() {
+		bootOutputStop <- true
+		<-bootOutputStop
+		close(bootOutputStop)
+	}
+
+	// Wait for ssh server to come up.
+	time.Sleep(5 * time.Second)
+	start := time.Now()
+	for {
+		c, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%v", inst.sshhost, inst.sshport), 1*time.Second)
+		if err == nil {
+			c.SetDeadline(time.Now().Add(1 * time.Second))
+			var tmp [1]byte
+			n, err := c.Read(tmp[:])
+			c.Close()
+			if err == nil && n > 0 {
+				break // ssh is up and responding
+			}
+			time.Sleep(3 * time.Second)
+		}
+		if time.Since(start) > 2*time.Minute {
+			done()
+			inst.merger.Wait()
+			return vmimpl.BootError{Title: "ssh server did not start", Output: bootOutput}
+		}
+	}
+	done()
+
+	return nil
+}
+
+func (inst *instance) Close() {
+	if out, err := inst.vmctl("stop", inst.vmIdent(), "-f"); err != nil {
+		if inst.debug {
+			log.Logf(0, "vmctl stop %s: %v: %s", inst.vmIdent(), err, string(out))
+		}
+	}
+
+	os.Remove(inst.image)
+}
+
+func (inst *instance) Forward(port int) (string, error) {
+	addr := fmt.Sprintf("100.64.%v.2:%v", inst.vmID, port)
+	return addr, nil
+}
+
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	vmDst := filepath.Join("/root", filepath.Base(hostSrc))
+	args := append(vmimpl.SCPArgs(inst.debug, inst.sshkey, inst.sshport), hostSrc, inst.sshuser+"@"+inst.sshhost+":"+vmDst)
+	cmd := osutil.Command("scp", args...)
+	if inst.debug {
+		log.Logf(0, "running command: scp %#v", args)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stdout
+	}
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+	done := make(chan bool)
+	go func() {
```

Comment:
Use `osutil.RunCmd` instead. That qemu code was very old and unnecessary complex.